src/lib/Libauth/gss/pbs_gss.c

```diff
@@ -0,0 +1,1184 @@
+		} else {
+			snprintf(gss_log_buffer, LOG_BUF_SIZE, "Entered encrypted communication with server %s", gss_extra->hostname);
+			GSS_LOG_DBG(gss_log_buffer);
+		}
+	}
+
+	return rc;
+}
+
+/** @brief
+ *	pbs_auth_encrypt_data - encrypt data based on given GSS context.
+ *
+ * @param[in] ctx - pointer to external auth context
+ * @param[in] data_in - clear text data
+ * @param[in] len_in - length of clear text data
+ * @param[out] data_out - encrypted data
+ * @param[out] len_out - length of encrypted data
+ *
+ * @return	int
+ * @retval	0 on success
+ * @retval	1 on error
+ */
+int
+pbs_auth_encrypt_data(void *ctx, void *data_in, size_t len_in, void **data_out, size_t *len_out)
+{
+	pbs_gss_extra_t *gss_extra = (pbs_gss_extra_t *)ctx;
+	OM_uint32 maj_stat;
+	OM_uint32 min_stat = 0;
+	gss_buffer_desc unwrapped;
+	gss_buffer_desc wrapped;
+	int conf_state = 0;
+
+	if (gss_extra == NULL) {
+		GSS_LOG_ERR("No GSS auth extra available");
+		return PBS_GSS_ERR_INTERNAL;
+	}
+
+	wrapped.length = 0;
+	wrapped.value = NULL;
+
+	unwrapped.length = len_in;
+	unwrapped.value  = data_in;
+
+	maj_stat = gss_wrap(&min_stat, gss_extra->gssctx, gss_extra->is_secure, GSS_C_QOP_DEFAULT, &unwrapped, &conf_state, &wrapped);
+
+	if (maj_stat != GSS_S_COMPLETE) {
+		GSS_LOG_STS("gss_wrap", maj_stat, min_stat);
+
+		maj_stat = gss_release_buffer(&min_stat, &wrapped);
+		if (maj_stat != GSS_S_COMPLETE) {
+			GSS_LOG_STS("gss_release_buffer", maj_stat, min_stat);
+			return PBS_GSS_ERR_INTERNAL;
+		}
+
+		return PBS_GSS_ERR_WRAP;
+	}
+
+	*len_out = wrapped.length;
+	*data_out = malloc(wrapped.length);
+	if (*data_out == NULL) {
+		GSS_LOG_ERR("malloc failure");
+		return PBS_GSS_ERR_INTERNAL;
+	}
+	memcpy(*data_out, wrapped.value, wrapped.length);
+
+	maj_stat = gss_release_buffer(&min_stat, &wrapped);
+	if (maj_stat != GSS_S_COMPLETE) {
+		GSS_LOG_STS("gss_release_buffer", maj_stat, min_stat);
+		return PBS_GSS_ERR_INTERNAL;
+	}
+
+	return PBS_GSS_OK;
+}
+
+/** @brief
+ *	pbs_auth_decrypt_data - decrypt data based on given GSS context.
+ *
+ * @param[in] ctx - pointer to external auth context
+ * @param[in] data_in - encrypted data
+ * @param[in] len_in - length of encrypted data
+ * @param[out] data_out - clear text data
+ * @param[out] len_out - length of clear text data
+ *
+ * @return	int
+ * @retval	0 on success
+ * @retval	1 on error
+ */
+int
+pbs_auth_decrypt_data(void *ctx, void *data_in, size_t len_in, void **data_out, size_t *len_out)
+{
+	pbs_gss_extra_t *gss_extra = (pbs_gss_extra_t *)ctx;
+	OM_uint32 maj_stat;
+	OM_uint32 min_stat = 0;
+	gss_buffer_desc wrapped;
+	gss_buffer_desc unwrapped;
+
+	if (gss_extra == NULL) {
+		GSS_LOG_ERR("No auth context available");
+		return PBS_GSS_ERR_INTERNAL;
+	}
+
+	if (len_in == 0) {
+		GSS_LOG_ERR("No data available to decrypt");
+		return PBS_GSS_ERR_INTERNAL;
+	}
+
+	if (gss_extra->is_secure == 0) {
+		GSS_LOG_ERR("wrapped data ready but auth context is not secure");
+		return PBS_GSS_ERR_INTERNAL;
```

Comment:
With GSS, we can wrap data but do not encrypt the data. Technically, it is not en error. The original idea behind this treated as the error was that we don't want to send credentials on an insecure line (only wrapped). Since we have PBS_ENCRYPT_MODE and we can disable encryption of everything anyway. It is a question whether this should be treated as a warning only... Just saying... I am not sure, and we can discuss it. I would slightly prefer still treat it as the error in PBS. Theoretically and in general, you could have clients using wrapped data but not encrypted.

Reply:
I also feel that we should treat as an error in PBS as sending cred in just wrapped (not encrypted) is not a good idea. If the user/admin doesn't care about the encryption part then it can disable it via PBS_ENCRYPT_MODE anyway. So I think if PBS_ENCRYPT_MODE is > 0 and GSS/Kerberos is not configured for encryption then we should treat as an error in PBS.
(Just FYI: If PBS_ENCRYPT_MODE is 0 (aka disabled) then PBS will never call pbs_auth_encrypt_data() and pbs_auth_decrypt_data() for any auth lib)

@subhasisb and others What's ur thought on this?